wireguard: Modularization of Agent

smagnani96 opened the pull request:
Each change has been carefully split in a separate logical commit to ease the review process.
Please refer to each commit and its description for further details.
I'd strongly suggest to review files owned by each group.

Major concerns:
* `wireguard: move local node to config param for updating EncryptionKey`: discussed and solved, see [below](https://github.com/cilium/cilium/pull/40360#discussion_r2185344257).
* `wireguard: rely on mtu reconciler job only for MTU settings`: please refer to commit description. This one is not strictly needed anymore (no more circular deps), but it feels right to keep. The link is created with default MTU, then the reconciler job will adjust it accordingly. No problems are expected for pods/host connections, given they'd rely on the already established routeMTU from the mtu subsystem.

Edit: there was a security threat discussion here https://github.com/cilium/cilium/pull/40360#discussion_r2222930079. The concerns arise from a previous implementation, but we noticed that there's an unrelated possibility to for an attacker to mess with the encryption by changing the key from k8s objects (CiliumEndpoint and CiliumNode). So to sum up, here we're not touching this anymore, but the threat is still available in Cilium.

Fixes: #25262

```release-note
Modularization of WireGuard Agent.
```

giorio94 commented on pkg/wireguard/agent/agent.go:
```diff
@@ -153,7 +151,20 @@ func newWireguardAgent(params params) *Agent {
+				job.OneShot("mtu-reconciler", agent.mtuReconciler),
+				job.OneShot("peer-reconciler", agent.peerReconciler),
+				job.OneShot("ipcache-listener", agent.ipsecListener),
+				job.OneShot("localnode-updater", agent.localnodeUpdater),
```
Drive-by comment. I don't think that this job is strictly guaranteed to be executed before that the LocalNode is first written to the CiliumNode/kvstore representation (even though it is likely to, easy to test adding a sleep). If that's the case, there would be the risk of incorrectly marking the local node as not needing encryption, causing packets to be sent out as plain text until it gets updated and reconciled again.

smagnani96 replied:
That was one of the concerns I had.
Maybe that part should be executed in the Start itself and not in a job.
Seeing also what we do for IPSec in the [Start](https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec/cell.go#L66-L68).

Thanks Marco for the feedback.

giorio94 replied:
Yeah, running it as part of the start hook makes the lifecycle more predictable. You should also make sure that the start hook is always executed before the nodediscovery is started, e.g., adding this cell as one of its dependencies (assuming it doesn't trigger a circular dependency, otherwise it becomes a bit more tricky).

smagnani96 replied:
>  You should also make sure that the start hook is always executed before the nodediscovery is started

IIUC, they're all daemon dependencies.
`NodeDiscovery` should start first, then `datapath.IPsecKeyCustodian` and `datapath.WireguardAgent` given they require `NodeDiscovery`.
However, the discovery itself is manually started [here](https://github.com/cilium/cilium/blob/main/daemon/cmd/daemon.go#L620-L622) in `newDaemon()`. At that point in time, both IPSec and WG should've set the needed bits.
So if I execute in the Start hook itself (and not in a job) we should be good to go.

giorio94 replied:
Yeah, it works because the discovery is started manually at the moment. Could you add an explicit comment to that node discovery method, so that we don't forget if we were ever to move that to a start hook?

smagnani96 replied:
Done.

smagnani96 committed:
wireguard: move IPSec+Wireguard enabled check to daemon

In this commit we move the current check whether both IPSec and Wireguard
are enabled out from the WireGuard agent to the newDaemon function.
Given upcoming changes, it is not a WireguardAgent responsibility to check
for this potential inconsistency, but rather is the daemon that should
detect and/or prevent potential conflicting configs.
This commit helps us towards the direction of cellifying the WireGuard
agent, having a plain simple constructor. Next commits will simplify that
even more removing possible syscall or nil returns.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: refactor needIPCacheEvents to private function

This commits refactors the previous needIPCacheEvents variable of the
WireGuard agent into a private helper function. There's no need to store
an additional variable, that is computed at creation time, given the
configuration is always accessible. This holds true also for upcoming
commits, where the agent will receive a private custom configuration.
This should help simplifying even more the agent creation process.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: refactor agent to Cell

This commit applies a simple refactor to the Wireguard agent Cell.
We define the parameters for initialization, that are accepted during the
cell creation. The Agent will be fully "cellified" in next commits.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: add daemon config as init parameter

This commit is a slow start for moving to using the global `option.Config`
to a local config. Here we accept the daemon config as an init parameter.
In subsequent commits, we will move to a custom local wireguard config
with a minimal set of values. But before doing this, we need to complete
cellifying the agent. Let's for now store a pointer to the daemon config.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: move client creation to Init

This commit moves the go wireguard client creation in the Init function.
This way, we go through a simple plain newAgent function with no
system calls or anything that can fail, slow down the startup, or return
a nil agent.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: move link deletion when disabled to agent Start

This commit, similarly to the previous one, aims at simplifying the
newAgent function to always return a non-nil value. We move
the wireguard link (cilium_wg0) deletion to the Start hook, for
those cases when wireguard is disabled.
At the same time, we protect the Start/Stop/Status public APIs with
the check on EnableWireguard. The other public APIs are for now already
used after checking EnableWireguard, and in subsequent commits they will
be moved internally to the agent. Let's not pollute commit history then.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: refactor peer-related APIs to private methods

This commit simply refactors unneeded public methods of the WireGuard
agent to private. This is possible given that these methods are internally
called from the agent itself upon node events.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: move local node to config param for updating EncryptionKey

Similarly as IPSec already does, let's move the LocalNode as a parameter
of the wireguard agent, and proactively modify the EncryptionKey once the
setup is completed with success. For now, let's attach these bits in the
deferred function inside the agent.Init(): in subsequent commits, once
the agent is fully cellified, we'll create apposite jobs to not hold the
Init/Start.
This commit also adds a note to the NodeDiscovery.StartDiscovery function:
we manually call it from `newDaemon()`, where both the WireGuard and IPSec
agent have already been initialized and started. Given they apply changes
to the local node, they must execute first BEFORE the `StartDiscovery()`,
otherwise it may lead to disruptions or leaked plain-text packets.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: add NodeManager to init parameter

Similarly as in the previous commits, let's add the NodeManager as init
parameter of the wireguard Cell. This helps subsequent commit, where we
move the execution of the Init function in the WireguardAgent cell start
itself. Again, let's add these bits in the deferred function in the
agent.Init(), which will be then moved to ad-hoc jobs in next commits.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: move Clustermesh and obsolete peer Restoration inside agent

Similarly as in previous commits, let's setup the RestoreFinished function
inside the wireguard agent itself, upon successfully finishing the
initialization. We add these bits in the deferred function in agent.Init
as in previous commits, which will be moved to ad-hoc jobs in next commits.
In addition, we move the check on clustermesh outside the restoreFinished
function: this way there's only one please (the go-routine) where we
wait for stuff, and leave `restoreFinished` to execute the needed action.

While doing that, we must preserve the sequence of action needed before
actually pruning stale peers. Before this commit, `startDaemon` was invoking
this GC routine after waiting for cache and IPCache synchronization.
This is needed to avoid accidentally pruning peers before the daemon actually
sees all local and remote nodes (and their allowed IPs). Let's preserve
this behavior by adding cacheStatus as a dependency to the WireGuard agent.
While moving the check, we do not wait for EndpointRestoration anymore
before garbage collecting old peers: the agent reconstructs its state by
looking at the nodeManager and IPCache entries.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: move IPCache to init param

This commit completes the set of changes needed so that every operation
on the Wireguard agent is registered from the agent itself upon starting
the cell. Here, we add IPCache as a dependency, and upon starting the agent,
in case of no errors, we autonomously register to all needed events.

Note: similarly for the IPSec key custodian, the WireGuard agent now
independently modifies the local node store upon starting. If we removed
the reference to the agent from the daemon params, the agent could
execute its Start routine after the LocalNode has already been written
to the CiliumNode/kvstore representation. This would cause problems,
given the agent (as for IPSec) sets the encryption key in the node object,
which is required for encryption to work (potential leakages observed
in node-to-node-encryption test otherwise).

we preserve the reference to the WireGuard agent in the daemon
params, so that we force it to start

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: refactor deferred functions to ad-hoc jobs

In previous commits, we moved all public APIs (RestoreFinished, Init) to
being privately called upon agent.Start. This allows us to refactor the
init function, where the deferred function executes a series of operations
in case the initialization process was successful. With this commit, we
move out some of these operations into ad-hoc jobs launched from the agent.Start().
In particular, we keep as jobs:

1. mtu-reconciler: no changes with respect to previous commits
2. peer-gc: this is the job waiting for kvstore and other components to
            sync before removing obsolete peers from the WireGuard device.

Among the functions kept in the agent.Start() and not as async jobs there are:
`localNode.Update(), `ipCache.AddListener()`, `nodeManager.Subscribe()`.
This is to align the behavior of the WireGuard agent to other subsystems
(e.g., IPSec calling localNode.Update() in its key manager Start method),
and to keep a clean and deterministic behavior of the agent.

For functions like localNode.Update(), we must ensure it runs in the
agent.Start function, which is required by the `newDaemon()`. In `newDaemon()`
we then call `nodeDiscovery.StartDiscovery()`, and at that point of time
the local node must be up-to-date with the WireGuard changes.

Refactoring into jobs also `ipCache.AddListener()` and `nodeManager.Subscribe()`
could've caused race conditions with the peer-gc job: the latter could
be executed before subscribing to node/ipcache events, causing the removal
of all restored entries.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: drop mtu on startup and rely only on mtu reconciler job

This commits remove the retrieval of the current MTU to be used in the
init function of the wireguard agent. This helps avoid getting stuck forever
in the agent Start, in case waiting for the watch channel of the mtu table.
Given we already have the reconciler job watching for MTU changes, let's
keep only that for this subsystem.

On agent restart, the link would not be recreated, therefore the previous
MTU in use would be kept, and the reconciler job would watch for changes.

On a new agent start, the link is created with the default MTU, and the
reconciler job would patch it accordingly. Pods/Host connections would already
use the correct RouteMTU retrieved from MTU subsystem (routeMTU), so
packets are not expected to fragment.

This commit is pure cleanup and doesn't provide additional support for
this patch series. In subsequent commits, we're going to export a WireGuard
agent and config via interfaces, so that other subsystems do not necessarily
need to rely on the original one (less methods exported, no cyclic deps).
This would not cause troubles here though (future WireGuard cell tests
would require more conditions).

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: move daemon flags to custom cell flags and config

This commits moves the `--enable-wireguard`, `wireguard-persistent-keepalive`,
`wireguard-track-all-ips-fallback` and `--node-encryption-opt-out-labels`
flags from the daemon config to the WireGuard cell, declaring also
user-provided flags and improving the final agent config structure.
In addition, it provides a private method to create the final configuration
given the user-provided flags and the daemon config.

In addition, this commits also includes the updated apis after running:
`make generate-k8s-api generate-api generate-health-api generate-hubble-api generate-operator-api generate-kvstoremesh-api`.

An extra effort was needed for NodeEncryptOptOutLabels, given prior to this
commit we were saving and retrieving them from the daemon config. With this
commit, those labels are stored in the WireGuard config, thus we need to
export them in the agent.Status function. In CLI tests, such info is not
retrieved anymore in `extractFeaturesFromRuntimeConfig`, but rather in
`extractFeaturesFromCiliumStatus` as for all the other features held in their packages.

The CLI change described above can handle upgrades/downgrades to/from other
Cilium versions:

* <  v1.19, NodeEncryptionOptOutLabels are stored in the DaemonConfig
* >= v1.19, NodeEncryptionOptOutLabels are stored in the Wireguard agent config

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: add cell tests

In previous commit, we fully cellified the WireGuard agent. While being
useful to have tests in each commits, it would require a lot of code changes
and insertions that would be later removed in subsequent commits.
For this reason, I opted for a final commit with the tests. Here we make
sure the Cell can successfully start, setup interface and client, and run
the parallel jobs upon finishing the startup process.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 commented on pkg/endpoint/endpoint_test.go:
```diff
@@ -252,7 +252,7 @@ func TestEndpointState(t *testing.T) {
 
 	model := newTestEndpointModel(100, StateWaitingForIdentity)
-	e, err := NewEndpointFromChangeModel(t.Context(), hivetest.Logger(t), nil, &MockEndpointBuildQueue{}, nil, s.orchestrator, nil, nil, nil, identitymanager.NewIDManager(logger), nil, nil, s.repo, testipcache.NewMockIPCache(), &FakeEndpointProxy{}, testidentity.NewMockIdentityAllocator(nil), ctmap.NewFakeGCRunner(), nil, model)
+	e, err := NewEndpointFromChangeModel(t.Context(), hivetest.Logger(t), nil, &MockEndpointBuildQueue{}, nil, s.orchestrator, nil, nil, nil, identitymanager.NewIDManager(logger), nil, nil, s.repo, testipcache.NewMockIPCache(), &FakeEndpointProxy{}, testidentity.NewMockIdentityAllocator(nil), ctmap.NewFakeGCRunner(), nil, model, nil)
```
```suggestion
	e, err := NewEndpointFromChangeModel(t.Context(), hivetest.Logger(t), nil, &MockEndpointBuildQueue{}, nil, s.orchestrator, nil, nil, nil, identitymanager.NewIDManager(logger), nil, nil, s.repo, testipcache.NewMockIPCache(), &FakeEndpointProxy{}, testidentity.NewMockIdentityAllocator(nil), ctmap.NewFakeGCRunner(), nil, model, &fakeTypes.WireguardAgent{})
```

Will solve this while addressing reviews.

smagnani96 commented on pkg/wireguard/agent/cell_test.go:
```diff
@@ -0,0 +1,314 @@
+// SPDX-License-Identifier: Apache-2.0
```
CI Build Commit taking too long to run and failing.
If it's essential that it passes or if reviewer believe I can squash some commits, please let me know.

gandro replied:
I vote for keep the individual commits around (assuming each commit itself builds) over pleasing CI and loosing the git history.

derailed commented on pkg/datapath/fake/types/wireguard.go:
```diff
@@ -0,0 +1,23 @@
+}
+
+func (fwa *WireguardAgent) Enabled() bool {
+	return fwa.EnableWireguard
```
Could add a guard here in case we're dealing with nil?

smagnani96 replied:
Will do in next force-push, trying to batch review feedback 🙏🏼

derailed commented on pkg/endpoint/api.go:
```diff
@@ -63,12 +63,12 @@ func (e *Endpoint) GetLabelsModel() (*models.LabelConfiguration, error) {
 
 // NewEndpointFromChangeModel creates a new endpoint from a request
-func NewEndpointFromChangeModel(ctx context.Context, logger *slog.Logger, dnsRulesAPI DNSRulesAPI, epBuildQueue EndpointBuildQueue, loader datapath.Loader, orchestrator datapath.Orchestrator, compilationLock datapath.CompilationLock, bandwidthManager datapath.BandwidthManager, ipTablesManager datapath.IptablesManager, identityManager identitymanager.IDManager, monitorAgent monitoragent.Agent, pol ...
+func NewEndpointFromChangeModel(ctx context.Context, logger *slog.Logger, dnsRulesAPI DNSRulesAPI, epBuildQueue EndpointBuildQueue, loader datapath.Loader, orchestrator datapath.Orchestrator, compilationLock datapath.CompilationLock, bandwidthManager datapath.BandwidthManager, ipTablesManager datapath.IptablesManager, identityManager identitymanager.IDManager, monitorAgent monitoragent.Agent, pol ...
```
Not part of this pr of course but this is adding to the smell...
Perhaps a endpoint config that could be passed down??

smagnani96 replied:
IIUC, you mean a similar solution to what we already do in the loader with `LocalNodeConfiguration` https://github.com/cilium/cilium/blob/main/pkg/datapath/types/node.go#L25-L32, right?
If that is the case yes, makes sense, but can be taken care in a != PR to carefully think the potential impact in up/downgrades (if any).

derailed reviewed (approved):
@smagnani96 Nice work Simone!

rgo3 reviewed (approved):
Nice work!

tklauser reviewed (approved):
ack for @cilium/fqdn

gandro commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -100,6 +100,11 @@ func NewNodeDiscovery(
+// NOTE: we manually call StartDiscovery() from newDaemon() after the Wireguard
+// and IPSec cell dependencies have been started/initialized, given they apply
+// changes on the local node. Be aware of this behavior in case of changes to
+// this function or cell.
```
informative (no change requested): Going forward, we might want to investigate the use the new fence mechanism for this kind of ordering dependency https://github.com/cilium/cilium/pull/39998

gandro commented on daemon/cmd/daemon_main.go:
```diff
@@ -1533,30 +1533,6 @@ func startDaemon(d *Daemon, restoredEndpoints *endpointRestoreState, cleaner *da
-				return
-			}
-
-			if err := params.WGAgent.RestoreFinished(d.clustermesh); err != nil {
```
So this call to `RestoreFinished` doesn't just require for KVstore things to be synced, it also requires local K8s nodes (`<-params.CacheStatus`))  and (where applicable) IPCache to be synced. Basically making sure we have seen all peers (i.e. remote nodes and their allowed IPs) before we prune stale peers in `restoreFinished`.

Historically, I think we placed this here so it's after `d.initRestore(restoredEndpoints, params.EndpointRegenerator)`, but I _think_ we don't really care about local endpoint restoration. Since we populate the WireGuard device from the node manager and IPCache, I think only those two are relevant.

By moving this into the `wireguard.Agent.Init` function which is issued from a start hook, I think we are now calling restore finished _before_ K8s caches are synced I believe (as the comment says, in CRD mode, the `WaitForKVStoreSync` returns immediately), which means we might not have discovered all nodes and/or IPCache IPs yet when `restoreFinished` runs.

smagnani96 replied:
Right, and thanks for pointing this out.

Would waiting for k8s nodes and ipcache to be synced before the `WaitForKVStoreSync` in the job work? Something like:

```go
		// peer-reconciler deletes obsolete peers.
		job.OneShot("peer-reconciler", func(ctx context.Context, _ cell.Health) error {

			// we wait here for params.cacheStatus
			if a.cacheStatus != nil {
				select {
				case <-ctx.Done():
					// fail job
				case <-a.cacheStatus:
				}
			}

			// we wait here for IPCache
			if err := a.ipCache.WaitForRevision(ctx, 1); err != nil {
			   // fail job
			}

			// we wait here for kvstore
			if err := a.nodeDiscovery.WaitForKVStoreSync(ctx); err != nil {
				return err
			}
			
			// we wait here for identity watcher for kvstore
			if err := a.ipIdentityWatcher.WaitForSync(ctx); err != nil {
				return err
			}

			// we finally cleanup obsolete peers.
			if err := a.restoreFinished(); err != nil {
				a.logger.Error("Failed to set up WireGuard peers", logfields.Error, err)
				return err
			}
			return nil
		}),
```

giorio94 replied:
Overall makes sense to me (the check on `cacheStatus != nil` is probably redundant, as [always initialized](https://github.com/cilium/cilium/blob/pr%2Fsmagnani96%2Fwg-modularization/pkg/k8s/synced/cell.go#L43-L45)). One possibility to clean that up a bit and get rid of a couple of dependencies could be letting the `NodeManager` to directly expose a readiness signal (having separate signals for CRD and kvstore is an easy footgun), but that's mostly unrelated to this PR. 

Another possibility could be directly depending on `regeneration.Fence` (which is the signal used to block endpoint regeneration). That would have the downside of additionally waiting on extra signals which are not of interest here, but that's most likely not a big deal. However, that one also currently suffers from not explicitly waiting for CiliumNode/ipcache synchronization (as implied by startup ordering), and fixing that would require some extra work as well. 

Yet another option could be using the a separate fence dedicated for wireguard, but again, that would require some more work to inject that into all dependencies, and may provide limited value in this context.

Regardless, I'd suggest moving the wait for clustermesh outside `restoreFinished`, for consistency.

gandro commented on pkg/node/manager/manager.go:
```diff
@@ -596,7 +596,8 @@ func (m *manager) nodeAddressHasEncryptKey() bool {
 // encryption key of the endpoint on that node are the same.
 func (m *manager) endpointEncryptionKey(n *nodeTypes.Node) ipcacheTypes.EncryptKey {
-	if m.conf.EnableWireguard {
+	// The public key is always set, even when opting-out from node-to-node encryption.
```
tiny nit: The commit message has a typo: `dependencyk`

gandro commented on pkg/wireguard/agent/cell_test.go:
```diff
@@ -0,0 +1,314 @@
+	testutils.PrivilegedTest(t)
+
+	// Create a temporary directory, used also by the agent for the private key.
+	testRunDir, err := os.MkdirTemp("", "cilium-wireguard-go-test")
```
nit: Consider using `t.TempDir()`

gandro reviewed (dismissed):
Appreciated the very readable commit history! :heart: Apologies for the late review. There is one major concern I have around `restoreFinished`. Other than that, this looks excellent

tommyp1ckles commented on pkg/wireguard/agent/cell_test.go:
```diff
@@ -0,0 +1,314 @@
+						NodeEncryptionOptOutLabels: nil,
+					}
+				},
+				func() tunnel.Config { return tunnel.Config{} },
```
nit: we could just define a output struct that composes `cell.Out` instead of having to define all these functions.

tommyp1ckles commented on pkg/wireguard/agent/cell_test.go:
```diff
@@ -0,0 +1,314 @@
+// SPDX-License-Identifier: Apache-2.0
```
Just a thought, these cell tests seem like they're a step towards Hive script tests. Maybe that's something to be looked at in the future.

tommyp1ckles reviewed (approved):
K8s changes look good, just a few small thoughts. Nice work!

giorio94 commented on pkg/wireguard/agent/agent.go:
```diff
@@ -87,26 +89,47 @@ type Agent struct {
+	}
 
+	agent := &Agent{
+		logger:   params.Logger.With(subsysLogAttr...),
```
Nit. The `.With(subsysLogAttr...)` is likely redundant, as the logger is already enriched with the name of the `cell.Module`.

giorio94 commented on pkg/wireguard/agent/agent.go:
```diff
@@ -143,14 +143,66 @@ func newWireguardAgent(params params) *Agent {
+		job.OneShot("nodemanager-subscribe", func(context.Context, cell.Health) error {
+			a.nodeManager.Subscribe(a)
+			return nil
+		}),
```
Nit. I'd suggest merging these two into a single job, given that they are both short lived. I'm honestly a bit on the fence on whether they should be just plainly executed as part of the start hook, to keep the logic as simple as possible. 

For instance, there's currently a theoretical race condition which could cause `restoreFinished` to be executed before subscribing to node/ipcache events, hence causing the removal of all restored entries. I reckon that it is unlikely that it will happen, but if the concern is the time taken by these functions, then the likelihood may not be even that low. 

One trick to reduce the number of updates may be also registering to the ipcache first (so that all updates are no-ops given that the corresponding nodes are unknown), and only then to the node manager.

giorio94 commented on pkg/node/manager/manager.go:
```diff
@@ -596,7 +596,8 @@ func (m *manager) nodeAddressHasEncryptKey() bool {
 // encryption key of the endpoint on that node are the same.
 func (m *manager) endpointEncryptionKey(n *nodeTypes.Node) ipcacheTypes.EncryptKey {
-	if m.conf.EnableWireguard {
+	// The public key is always set, even when opting-out from node-to-node encryption.
```
Nit. The commit message is a bit misleading, as it refers to "local node", even though the node here will be a remote one from the point of view of the local node/agent. 

Additionally, this is also a behavioral change, as the decision on whether to encrypt is shifted from a local configuration (i.e., the flag value) to something that depends on a external information (i.e., whether the public key is advertised), which has implications in terms of the security model. That said, this only controls the encryption of the health/ingress endpoints, the same already applies to the IPSec case, and to all the other pods, so this looks totally fine to me.

smagnani96 replied:
Will adjust commit description.

I did this change to avoid circular deps, and haven't thought about the security model.
AFAIU, the public key field of a node is updated before it gets advertised, so I'd say this check eventually ends up to being true. But I do understand the conceptual shift... how should we proceed here?

giorio94 replied:
> AFAIU, the public key field of a node is updated before it gets advertised, so I'd say this check eventually ends up to being true.

Yep, it is more that a potential attacker able to control the CiliumNode would now be able to disable encryption for these endpoints. 

I was double-checking the behavior for standard endpoints:

* When running in CRD mode, the encryption key is defaulted to the local one, and overridden if specified in the CiliumEndpoint \[1]. I haven't checked, but I wouldn't be surprised if it was possible to override it with 0 (to disable encryption) if set in the resource.  
* When running in kvstore mode, the encryption key part of the etcd entry is simply propagated \[2]. 

So unless I'm missing something, I'd say that your change is not allowing anything that wouldn't already be possible for standard endpoints under the same attacker model. It would be good to double-check with a manual test though, just to make sure that there are not other safeguards in place.

That said, if you don't want to change anything, you could break the circular dependency by providing a separate `WireguardEnabled` type (based on the configuration only), which can then be consumed by all the other subsystems wishing to know whether it is enabled.

\[1]: https://github.com/cilium/cilium/blob/pr%2Fsmagnani96%2Fwg-modularization/pkg/k8s/watchers/cilium_endpoint.go#L162-L171
\[2]: https://github.com/cilium/cilium/blob/pr%2Fsmagnani96%2Fwg-modularization/pkg/ipcache/kvstore.go#L426-L429

smagnani96 replied:
Marco I just verified what you just said. Of course this does not apply in Overlay mode where we always encrypt traffic. I'm testing locally in CRD mode:

1. Changing `EncryptKey: 0` in a CiliumEndpoint leads to plain-text traffic (pod-to-pod-encryption-v2 test)
2. Changing `EncryptKey: 0` in a CiliumNode leads to plain-text traffic (node-to-node-encryption test)
3. Changing annotation WG pubKey in CiliumNode (my PR) doesn't lead to plain-text traffic (node-to-node-encryption test), unless also (2).

Based on this, what's the best way to proceed with the PR?
I've addressed all the other feedback locally, wanna tackle this before force-pushing again 🙏🏼

giorio94 replied:
> Changing annotation WG pubKey in CiliumNode (my PR) doesn't lead to plain-text traffic (node-to-node-encryption test),

That would only affect ingress/health traffic (which is not validated by any CLI encryption test AFAIK), as those are the only retrieving the key via the function you modified.

> Based on this, what's the best way to proceed with the PR?

Given that encryption for practically all other endpoints is already controlled via CiliumNode/CiliumEndpoint, I'd say this doesn't change anything from a security standpoint, so your current approach seems totally fine IMO.

smagnani96 replied:
Many many thanks Marco again, also for the offline discussion.
I'll now push the new code, where I've exported both the WireGuard `Agent` and the cell `Config` as interfaces.
This has the advantage that we don't have to change the behavior discussed above.
That said, the security analysis was super useful, and we've verified that changing the encryption key in k8s objects will cause packets to be leaked plain text).

giorio94 commented on pkg/wireguard/agent/agent_test.go:
```diff
@@ -283,7 +285,7 @@ func TestAgent_PeerConfig(t *testing.T) {
 	} {
 		t.Run(c.Name, func(t *testing.T) {
-			wgAgent, ipCache := newTestAgent(t.Context(), hivetest.Logger(t), newFakeWgClient(), c.toAgentConfig())
+			wgAgent, ipCache := newTestAgent(t.Context(), hivetest.Logger(t, hivetest.LogLevel(slog.LevelError)), newFakeWgClient(), c.toAgentConfig())
```
Nit. Is this change intentional? Same below.

smagnani96 replied:
Yep, it is intentional so that we don't get unneeded logs during tests.
Should I split into a separate commit?

giorio94 replied:
I think (but I may be wrong) that the typical approach in the rest of the codebase is to keep the default log level, as `go test` automatically displays them only in case of failure, or if `--test.v` is specified. But I'm not strongly opposed to changing it here.

smagnani96 replied:
You're not wrong at all actually. I was simply running with `-v` all the times 🤯 
Dropping this bits.

giorio94 commented on pkg/wireguard/agent/cell.go:
```diff
@@ -5,11 +5,61 @@ package agent
+		EnableIPv6:                 dc.EnableIPv6,
+		TunnelingEnabled:           dc.TunnelingEnabled(),
+		EncryptNode:                dc.EncryptNode,
+		NodeEncryptionOptOutLabels: dc.NodeEncryptionOptOutLabels,
```
Nit. Any reason for not moving `NodeEncryptionOptOutLabels` as well to the set of local configurations? From a quick look it seems accessed inside the wireguard subsystem only (and leftover usage in tests).

smagnani96 replied:
While checking what else value could go into a WG-only config, I missed this one.
Will do 👍🏼

giorio94 commented on pkg/wireguard/agent/cell_test.go:
```diff
@@ -0,0 +1,314 @@
+				ipcache.NewIPIdentitySynchronizer,
+				func(_ *statedb.DB, devices statedb.RWTable[*tables.Device]) statedb.Table[*tables.Device] {
+					return devices
+				},
```
Nit. Shorter version: `statedb.RWTable[*tables.Device].ToTable,`

giorio94 commented on pkg/wireguard/agent/cell_test.go:
```diff
@@ -0,0 +1,314 @@
+			require.NoError(t, hive.Start(log, ctx))
+
+			// 2. Ensure the wireguard Agent is enabled.
+			assert.True(t, wgAgent.Enabled())
```
Nit. I would avoid mixing `require` and `assert` (except inside the `EventuallyWithT` blocks, where you need to use `assert`), for consistency and to avoid surprises.

giorio94 commented on pkg/wireguard/agent/cell_test.go:
```diff
@@ -0,0 +1,314 @@
+			assert.False(t, wgAgent.Enabled())
+
+			// 3. Ensure the link is deleted.
+			_, err = safenetlink.LinkByName(types.IfaceName)
+			require.Error(t, err)
```
If you want to assert that the link got deleted, it is much better to create it yourself, rather than relying on leftover state from a previous test, as that is not guaranteed to have been executed.

giorio94 commented on pkg/wireguard/agent/agent.go:
```diff
@@ -143,14 +143,66 @@ func newWireguardAgent(params params) *Agent {
+			return nil
+		}),
+		// peer-reconciler deletes obsolete peers.
+		job.OneShot("peer-reconciler", func(ctx context.Context, _ cell.Health) error {
```
Nit. This job name is slightly misleading (reconciler seems to hit to a long running process). Maybe `gc`?

I also wonder if it could make sense configuring e.g., three retries, considering that `restoreFinished` can return errors.

giorio94 reviewed (commented):
Amazing work, thanks for the very readable commit history, and sorry for the delay. A few comments inline, mostly nits.

smagnani96 committed:
wireguard: create common agent interface

This commit creates a common interface type for the wireguard
agent to avoid circular dependencies and to limit the set of operation
that all the other cells can call on the WireGuard agent. We now add two:

1. Enabled():    returns true/false whether the agent is enabled or not;
2. Status(bool): returns a *models.WireguardStatus representing the status
                 of the agent.

Given this change, the commits modifies all the other cells/subsystems
previously relying on the *Agent accordingly to the new wgTypes.WireguardAgent.
Among the subsystems in discussions there are: debug/api and status.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

smagnani96 committed:
wireguard: create common config interface

This commit creates a common config interface type for the wireguard
config to avoid circular dependencies and to limit the set of operation
that all the other cells can call on the config itself. This is particularly
useful in all those cells/subsystems that needs to check whether WireGuard
is enabled, but don't want (or cannot due to Start priorities or circular
dependencies) to depend on the whole agent. The new interface exports the
`Enabled()` method, which returns true when the agent is enabled.

Given this change, the commits modifies all the other cells/subsystems
previously relying on the daemon.Option.EnableWireguard accordingly to the
new wgTypes.WireguardConfig.Enabled(). Among the subsystems in discussions
there are: metrics, endpoint, orchestrator, node, loader.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

giorio94 commented on pkg/wireguard/agent/agent.go:
```diff
@@ -309,16 +363,74 @@ func (a *Agent) mtuReconciler(ctx context.Context, health cell.Health) error {
+// peerGarbageCollector is a job that removes obsoletes peers from the WireGuard device.
+// If an error is encountered, the job will retry with exponential backoff.
+func (a *Agent) peerGarbageCollector(ctx context.Context, health cell.Health) error {
+	retryTimer := backoff.Exponential{Logger: a.logger, Min: 100 * time.Millisecond, Max: 1 * time.Minute}
```
Nit. The only reason why the different functions return an error is because the context got canceled (that is, the agent is terminating), so this logic can be simplified a bit. And one-shot jobs already support a [retry mechanism](https://github.com/cilium/cilium/blob/pr%2Fsmagnani96%2Fwg-modularization/vendor/github.com/cilium/hive/job/oneshot.go#L70-L78), no need for a custom one.

smagnani96 replied:
Given I haven't run CI yet, I'll update the PR with this suggestions.
Would it make sense to simplify also the other job `peer-reconciler` in a specific commit (remove custom retry mechanism in favor of WithRetry), or that would drive the attention out of the scope of the PR?

giorio94 replied:
> Would it make sense to simplify also the other job [...]

As you prefer, could be a follow-up as well given that all reviews are already in.

smagnani96 replied:
Done (return nil in all the other cases), I've also moved the inline comments into the function for readability.
Let me know 😃

giorio94 replied:
Looks great, thanks!

giorio94 reviewed (approved):
Amazing, thanks!
